ddtrace/debugging/_probe/registry.py

```diff
@@ -0,0 +1,185 @@
+        return self.probe.activate()
+
+    def deactivate(self):
+        # type: () -> None
+        return self.probe.deactivate()
+
+    def set_installed(self):
+        # type: () -> None
+        self.installed = True
+
+    def set_exc_info(self, exc_info):
+        # type: (ExcInfoType) -> None
+        self.exc_info = exc_info
+
+    def set_message(self, message):
+        # type: (str) -> None
+        self.message = message
+
+
+def _get_probe_location(probe):
+    # type: (Probe) -> Optional[str]
+    if isinstance(probe, LineProbe):
+        return probe.source_file
+    elif isinstance(probe, FunctionProbe):
+        return probe.module
+    else:
+        raise ValueError("Unsupported probe type: {}".format(type(probe)))
+
+
+class ProbeRegistry(dict):
+    """Keep track of all the registered probes.
+
+    New probes are also registered as pending, on a location basis, until they
+    are processed (e.g. installed, generally by some import hook). Pending
+    probes can be retrieved with the ``get_pending`` method.
+    """
+
+    def __init__(self, status_logger, *args, **kwargs):
+        # type: (ProbeStatusLogger, Any, Any) -> None
+        """Initialize the probe registry."""
+        super(ProbeRegistry, self).__init__(*args, **kwargs)
+        self.logger = status_logger
+
+        # Used to keep track of probes pending installation
+        self._pending = defaultdict(list)  # type: Dict[str, List[Probe]]
+
+        self._lock = forksafe.RLock()
+
+    def register(self, *probes):
+        # type: (Probe) -> None
+        """Register a probe."""
+        with self._lock:
+            for probe in probes:
+                self[probe.probe_id] = ProbeRegistryEntry(probe)
+
+                location = _get_probe_location(probe)
+                if location is None:
+                    self.set_error(probe, "Unable to resolve location information for probe {}".format(probe.probe_id))
+                    continue
+
+                self._pending[location].append(probe)
+
+                self.logger.received(probe)
+
+    def set_installed(self, probe):
+        # type: (Probe) -> None
+        """Set the installed flag for a probe."""
+        with self._lock:
+            self[probe.probe_id].set_installed()
+
+            # No longer pending
+            self._remove_pending(probe)
+
+            self.logger.installed(probe)
+
+    def set_exc_info(self, probe, exc_info):
+        # type: (Probe, ExcInfoType) -> None
+        """Set the installed flag for a probe."""
+        with self._lock:
+            self[probe.probe_id].set_exc_info(exc_info)
+            self.logger.error(probe, exc_info=exc_info)
+
+    def set_error(self, probe, message):
+        # type: (Probe, str) -> None
+        """Set the error message for a probe."""
+        with self._lock:
+            self[probe.probe_id].set_message(message)
+            self.logger.error(probe, message)
+
+    def log_probes_status(self):
```

Comment:
Where is this method used? Or is the plan to use it in future work?

Reply:
Yep, this will be called by stuff that comes later.